### PR TITLE
Upgrade to omniauth 2.0

### PR DESCRIPTION
**Pull request by greatestape**
### What this PR does

* Makes shopify_app compatible with omniauth 2.x

### Reviewer's guide to testing

* Create a new app or install or
* Install the local checkout of the gem and `bundle install`

### Things to focus on

* [x] Apps created from scratch have omniauth 2.x as a dependency out of the box
* [x] Existing app can update to the new version of `shopify_app` and then update to omniauth 2.x
* [x] Oauth works
  * Actions
    * [x] App install
    * [x] Requesting new scopes
  * Scenarios
    * [x] In Chrome
    * [x] In Safari
    * [x] In an app using Session Tokens
    * [x] In an app using 3P session cookies
    * [x] In an app using only offline tokens
    * [x] In an app using offline and online tokens

### Checklist

Before submitting the PR, please consider if any of the following are needed:

- [x] Update `CHANGELOG.md` if the changes would impact users
- [x] Update `README.md`, if appropriate.
- [x] Update any relevant pages in `/docs`, if necessary
- [x] For security fixes, the [Disclosure Policy](https://github.com/Shopify/shopify_app/blob/master/SECURITY.md#disclosure-policy) must be followed.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,6 @@
 Unreleased
 ----------
+* Support omniauth 2.x
```
Comment (andyw8) [inline CHANGELOG.md:3]: I think we'll need some instructions here for people who are upgrading?

File: test/shopify_app/session/jwt_test.rb [hunk 2]
```diff
@@ -50,7 +50,7 @@ class JWTTest < ActiveSupport::TestCase
 
     test "#shopify_domain and #shopify_user_id are nil if the jwt is unsigned" do
       # The library expects there to be a 3rd segment if we want to verify signature
-      expect_jwt_error(::JWT::DecodeError, 'Not enough or too many segments')
+      expect_jwt_error(::JWT::IncorrectAlgorithm, 'Expected a different algorithm')
```
Comment (greatestape) [inline test/shopify_app/session/jwt_test.rb:53]: I wish the upgrade of `jwt` and this test change could be isolated in their own commit. If I ran across this change later and found that it was part of a commit upgrading omniauth I would be stumped. I realize it might be hard to tease apart the jwt upgrade from the rest of the upgrades, but I just wanted to put my hopes and dreams out there in the world.

File: lib/shopify_app/controller_concerns/login_protection.rb [hunk 3]
```diff
@@ -223,7 +223,6 @@ def sanitized_params
     end
 
     def return_address
-      return base_return_address unless ShopifyApp.configuration.allow_jwt_authentication
```
Comment (ragalie) [inline lib/shopify_app/controller_concerns/login_protection.rb:226]: What's the rationale behind this change?

File: test/shopify_app/session/jwt_test.rb (hunk 2 above)
Comment (rezaansyed) [reply test/shopify_app/session/jwt_test.rb:53]: Yeah I'm trying to see if we can avoid it but because we're allowing patch releases of `jwt` and the bump from `2.2.2` to `2.2.3` included this change, we may have to support it.

File: app/views/shopify_app/shared/post_redirect_to_auth_shopify.html.erb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  <title>Redirecting…</title>
+  <script>
+    function redirect() {
+      var form = document.getElementById("redirect-form");
+      if (form) {
+        form.submit();
+      }
+    }
+    window.addEventListener("load", redirect);
+  </script>
```
[the remark this answers is not in the document]
Comment (rezaansyed) [reply app/views/shopify_app/shared/post_redirect_to_auth_shopify.html.erb:16]: I think listening for `DOMContentLoaded` event makes sense. With regards to moving this to the bottom of `<body>`, not sure if that approach is required. @hannachen do you have any thoughts on this?

File: lib/shopify_app/controller_concerns/login_protection.rb (hunk 3 above)
Comment (rezaansyed) [reply lib/shopify_app/controller_concerns/login_protection.rb:226]: the return address previously didn't have the `shop` and `host` params when redirecting to the app after OAuth. Because of that, we're not able to use those params to load App Bridge to embed the app inside of an iframe. We can simply pass in the params in the return address now.

Comment (ragalie) [reply lib/shopify_app/controller_concerns/login_protection.rb:226]: Ok, can you help me understand how this is related to the omniauth upgrade?

Comment (rezaansyed) [reply lib/shopify_app/controller_concerns/login_protection.rb:226]: This isn't actually. I've gone ahead and created a [new PR](https://github.com/Shopify/shopify_app/pull/1259) for this. Could I get a review on that?

File: test/shopify_app/session/jwt_test.rb (hunk 2 above)
Comment (ragalie) [reply test/shopify_app/session/jwt_test.rb:53]: The gemspec still allows `2.2.2`, which I don't think will work properly after this change? https://github.com/Shopify/shopify_app/pull/1258/files#diff-d72fff1e948b092ced84db03fa37f7c2aa1774033d3aa250d41d847efb2a2264R21

You'll need to require `>= 2.2.3` for this change to be safe, I think.

File: app/views/shopify_app/shared/post_redirect_to_auth_shopify.html.erb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="utf-8" />
+  <meta name="viewport" content="width=device-width, initial-scale=1" />
+  <base target="_top">
+  <title>Redirecting…</title>
+  <script>
+    function redirect() {
```
[the remark this answers is not in the document]
Comment (rezaansyed) [reply app/views/shopify_app/shared/post_redirect_to_auth_shopify.html.erb:9]: Done.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,7 @@
...
```
Comment (andyw8) [inline CHANGELOG.md:3]:
```suggestion
* Support OmniAuth 2.x 
```

**Review by alexaitken — APPROVED** [review]
LGTM

Tested the following flows and configuration combinations
- Setup
  - New app
  - Upgrade existing app
- Configurations   
  - In an app using Session Tokens
  - In an app using 3P session cookies
  - In an app using only offline tokens
- Actions
     - App install
     - Requesting new scopes
- Browsers
   -  In latest Chrome
   -  In latest Safari
   -  In latest Firefox
